### PR TITLE
[wasm] optimize out 'is this method MulticastDelegate.Invoke' checks in aot->interp transitions

### PR DESCRIPTION
**Pull request by kg**
Right now interp_entry and jiterpreter's equivalent need to check for every call whether the call target is `MulticastDelegate.Invoke`. While this sounds inexpensive, in browser-bench's JSON suite that check by itself was ~0.5% of cpu time according to Chrome's profiler. This PR updates InterpMethod creation to store a flag for this so we can just check the flag during interp_entry to avoid the wasted work.

### CODE REVIEW

**Review by vargaz — APPROVED** [review]

**Comment by radical** [comment]
Fix for the STJ failures was just merged - https://github.com/dotnet/runtime/pull/80186 .

**Comment by kg** [comment]
The S.T.J AOT failures looked relevant, but I can't reproduce them so I'm bumping to latest upstream to re-run.

**Comment by lewing** [comment]
the wbt failures look like https://github.com/dotnet/runtime/pull/80241

File: src/mono/mono/mini/interp/jiterpreter.c [hunk 1]
```diff
@@ -575,7 +575,7 @@ mono_jiterp_interp_entry_prologue (JiterpEntryData *data, void *this_arg)
 	method = rmethod->method;
 	jiterp_assert(method);
 
-	if (mono_interp_is_method_multicastdelegate_invoke(method)) {
+	if (data->header.can_be_invoke && mono_interp_is_method_multicastdelegate_invoke(method)) {
```
Comment (lewing) [inline src/mono/mono/mini/interp/jiterpreter.c:578]: you've got a byte, is there some reason you can't store the result of the check once it has run too?

Comment (kg) [reply src/mono/mono/mini/interp/jiterpreter.c:578]: Great point, I'll do that.

Comment (kg) [reply src/mono/mono/mini/interp/jiterpreter.c:578]: Actually, I'm not certain this value is guaranteed to be the same on every call? @vargaz @BrzVlad do either of you know whether the 'is multicastdelegate invoke' state when entering the interp could change from call to call? It seems to me like it should be constant for a given entry wrapper.

Comment (lewing) [reply src/mono/mono/mini/interp/jiterpreter.c:578]: https://github.com/dotnet/runtime/blob/db28821852151f62fb81d83d0fd1c8de9e0e5640/src/mono/mono/mini/interp/interp.c#L2165 appears to be the logic

Comment (BrzVlad) [reply src/mono/mono/mini/interp/jiterpreter.c:578]: Aren't interp entry wrappers generated per method ? So we know at jiterp compile time if the method is multicast delegate invoke. Which means we could remove the checks for this altogether from interp entry path and generate a separate call to obtain the delegate wrapper when we know we are calling such a method. Or maybe not bother at all with compiling jit wrappers for these methods. Are they really that common ?

Comment (kg) [reply src/mono/mono/mini/interp/jiterpreter.c:578]: My understanding is that they're per-method so we know at jiterp compile time, yeah. I just wasn't 100% certain that I wasn't missing a code path where they could get shared. I don't know how to identify whether they're actually common in production software or not... I suspect they're going to get hit a lot in System.Text.Json, I'll see if I can verify that.

Comment (vargaz) [reply src/mono/mono/mini/interp/jiterpreter.c]: I can be stored in InterpMethod.

Comment (kg) [reply src/mono/mono/mini/interp/jiterpreter.c]: Verified that storing a flag in InterpMethod works and that the S.T.J tests hit this as I expected. So the optimization applies to both regular AOT->interp transitions and to the jiterpreter now. Vlad suggested generating optimized wrappers that skip the check entirely, which I'll look into doing.

File: src/mono/mono/mini/interp/interp.c [hunk 2]
```diff
@@ -2162,12 +2164,11 @@ typedef struct {
...
-is_method_multicastdelegate_invoke (MonoMethod *method)
+is_imethod_multicastdelegate_invoke (InterpMethod *imethod)
 {
-	gboolean expected = method->is_invoke;
-	gboolean actual = (m_class_get_parent (method->klass) == mono_defaults.multicastdelegate_class) && !strcmp (method->name, "Invoke");
-	assert (actual == expected)
-	return actual;
+	if (!imethod->is_invoke)
+		return FALSE;
+	return (m_class_get_parent (imethod->method->klass) == mono_defaults.multicastdelegate_class);
```
Comment (vargaz) [inline src/mono/mono/mini/interp/interp.c:2171]: Can't is_invoke be set based on this ?

Comment (kg) [reply src/mono/mono/mini/interp/interp.c:2171]: Is multicastdelegate_class always going to be available up front? Couldn't we be running cctors in the interpreter while we initialize types?

Comment (vargaz) [reply src/mono/mono/mini/interp/interp.c:2171]: mono_defaults is initialized very early.

Comment (kg) [reply src/mono/mono/mini/interp/interp.c:2171]: Will test it.

Comment (kg) [reply src/mono/mono/mini/interp/interp.c]: It works 👍

File: src/mono/mono/mini/interp/jiterpreter.c [hunk 3]
```diff
@@ -572,23 +571,21 @@ mono_jiterp_interp_entry_prologue (JiterpEntryData *data, void *this_arg)
...
-			method = cache.delegate_invoke;
 			data->header.rmethod = rmethod = cache.delegate_invoke_rmethod;
 		} else {
 			/*
 			* This happens when AOT code for the invoke wrapper is not found.
 			* Have to replace the method with the wrapper here, since the wrapper depends on the delegate.
 			*/
 			MonoDelegate *del = (MonoDelegate*)this_arg;
-			method = mono_marshal_get_delegate_invoke (method, del);
+			MonoMethod *method = mono_marshal_get_delegate_invoke (method, del);
```
Comment (lewing) [inline src/mono/mono/mini/interp/jiterpreter.c:588]:
```suggestion
			MonoMethod *method = mono_marshal_get_delegate_invoke (rmethod->method, del);
```

**Pull request merged**